More structs and a generic client

radu-matei opened the pull request:
This PR adds events, jobs, and workers structs, and implements a generic client to be used by specific clients.

Commit:
Add structs for events, jobs, and workers.

Signed-off-by: Radu M <root@radu.sh>

Commit:
Add generic client and use it in the projects client

Signed-off-by: Radu M <root@radu.sh>

Commit:
Additional constraints to generic client

Signed-off-by: Radu M <root@radu.sh>

Commit:
Update client usage in projects client

Signed-off-by: Radu M <root@radu.sh>

Commit:
Add WIP events client

Signed-off-by: Radu M <root@radu.sh>

Commit:
Add WIP events client

Signed-off-by: Radu M <root@radu.sh>

radu-matei commented on src/client.rs:
```diff
@@ -0,0 +1,208 @@
+        Ok(list)
+    }
+
+    // Specific client might have selectors for listing clients.
```
_specific clients_

radu-matei commented on src/client.rs:
```diff
@@ -0,0 +1,208 @@
+    }
+}
+
+// #[async_trait::async_trait]
```
Remove this, since we are no longer using an async trait.

radu-matei commented on src/client.rs:
```diff
@@ -0,0 +1,208 @@
+use serde_with::*;
+
+pub struct Client {
+    pub rest: rest::Client,
```
Naming two separate structs `Client`, even if both are internal, is clunky.
I there is no need for both of them.

radu-matei commented on src/client.rs:
```diff
@@ -0,0 +1,208 @@
+impl Client {
+    pub async fn get<
+        T: Serialize + DeserializeOwned + Sized,
+        U: Serialize + DeserializeOwned + Sized,
```
There is no need for two type parameters for a `get` operation.

krancour commented:
@radu-matei I love watching you review your own PR. 😂 

Sorry I can't be more help with reviewing Rust code.

Commit:
Keep a single generic `Client` struct

This commit merges the `rest::Client` and `client::Client` structs into
`client::Client`, and updates the usage across the existing clients.

Signed-off-by: Radu M <root@radu.sh>

Commit:
Remove deleted module from src/lib.rs

Signed-off-by: Radu M <root@radu.sh>
